File: GasStation.java
/**
 * Created by yhf on 15/4/22.
 */

/*
There are N gas stations along a circular route, where the amount of gas at station i is gas[i].

You have a car with an unlimited gas tank and it costs cost[i] of gas to travel from station i to its next station (i+1).
You begin the journey with an empty tank at one of the gas stations.

Return the starting gas station's index if you can travel around the circuit once, otherwise return -1.

Note:
The solution is guaranteed to be unique.
*/

/*
* 1） total = ∑ (gas[i] - cost[i]), 0 <= i < N;
*     如果total < 0，说明根本不可能绕一圈，返回-1；
*     如果total >= 0，说明至少存在一个start的点。
* 2） 计从start开始的 sum = ∑(gas-cost)，一旦sum<0，
*     就跳到下一个start，并将sum清零
* 3） 如果 i 到 j 不是一个解，那么start肯定不在 (i, j]。
*     因为如果假设k是一个满足条件的start，且 i < k <= j，
*     那么k到j的sum>0，那么i到k的sum<0，在k就应该停止了
* 4） 从2），3）知，由跳到下一个start时，直接跳到j+1即可。
* 5） 因为如果total>=0，start点肯定存在。当j到达N-1时，
*     因为已经证明此时的i之前的点都不是start点，
*     所以这时的i就是start点。由于start点唯一，所以循环结束。
* */

public class GasStation {
    public static int canCompleteCircuit(int[] gas, int[] cost) {
        int len = gas.length;
        int total = 0, sum = 0, start = 0;
        for (int i = 0; i < len; i++) {
            total += gas[i] - cost[i];
            sum += gas[i] - cost[i];
            if (sum < 0) {
                start = i + 1;
                sum = 0;
            }
        }
        return total >= 0? start : -1;
    }

    public static void main(String[] args) {
        int[] gas = {2, 4};
        int[] cost = {3, 4};
        int start = canCompleteCircuit(gas, cost);
    }
}

File: Leetcode-Python/GasStation.py
"""
There are N gas stations along a circular route, where the amount of gas at station i is gas[i].

You have a car with an unlimited gas tank and it costs cost[i] of gas to travel from station i to its next
station (i+1). You begin the journey with an empty tank at one of the gas stations.

Return the starting gas station's index if you can travel around the circuit once, otherwise return -1.

Note:
The solution is guaranteed to be unique.

1) global_left = sum(gas[i] - cost[i]), it is the total gas left after a round.
   if global_left < 0, means it cannot complete a circuit, returns -1.
2) suppose (i, j] is an interval where you cannot get from i to j. Then any k that i < k <= j
   cannot be a start point. Because if k is a start point, you must can get from k to j, which
   means sum(k to j) >= 0, then sum(from i to k) < 0, then you should be stop at k rather than j.
3) So, if you stop at j, update the start point at j + 1.
4) If no point in [0, len-1) is start point, then the last one must be. Because global_left >= 0,
   then there must be a start point.
5) We can optimize our solution in one-pass.
"""


class Solution:
    # @param {integer[]} gas
    # @param {integer[]} cost
    # @return {integer}
    def canCompleteCircuit(self, gas, cost):
        gas_left = global_left = 0
        start = 0
        for idx, (g, c) in enumerate(zip(gas, cost)):
            gas_left += g - c
            global_left += g - c
            if gas_left < 0:
                start = idx + 1
                gas_left = 0
        return start if global_left >= 0 else -1
